### PR TITLE
Machine learning callbacks

### PR DESCRIPTION
**Pull request by maffettone**
Construct a generic callback that interfaces the document stream with a passive machine learning agent, using the `tell`-`report` interface. 

## Description

The callback:
- focuses on passive ML, that is, agents that respond to but do not guide experiments. 
- consumes an arbitrary number of agents which are each required to have a `tell` and `report` method. 
- will track a specific stream (default "primary"), and requires explicit specification of independent and dependent variables.
- does not use "hints" as in the best effort callback.
- calls `agent.tell()` on a per-event basis
- calls `agent.report()` on a per-event and/or a per-stop basis. 

The documentation for callbacks was extended to include a ML section. 

## Motivation and Context

This follows on from reviewer and user feedback regarding [this paper](https://arxiv.org/abs/2201.03550), 
and the paradigms outlined there and in [bluesky-adaptive](https://blueskyproject.io/bluesky-adaptive/) and its respective [tutorials](https://github.com/bluesky/tutorials/blob/main/Adaptive%20RL%20Sampling/Adaptive%20Sampling.ipynb).

## How Has This Been Tested?

A full set of unit tests has been developed along side of the callback to ensure expected functionality.

### CODE REVIEW

File: bluesky/callbacks/ml.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        dependent_key: str,
+        report_on_event: bool = True,
+        report_on_stop: bool = True,
+        stream_name: str = "primary",
+    ):
+        super().__init__()
+        self.agents = list()
+        for agent in agents:
+            if self.agent_validation(agent):
+                self.agents.append(agent)
```
Comment (stan-dot) [inline bluesky/callbacks/ml.py:59]: could we just have a filter here?

```python

self.agents.extend(filter(lambda agent: isinstance(agent, Agent), agents))
```

**Comment by stan-dot** [comment]
to review in October

**Comment by stan-dot** [comment]
possibly move as issue into bluesky-adaptive

**Comment by maffettone** [comment]
This can be closed, as superseded by the bluesky-adaptive work.

**Pull request closed without merging**